### PR TITLE
Jvenstad/expose property info in application api

### PR DESCRIPTION
**Pull request by jonmv**
@mpolden please review. 

Replaces https://github.com/vespa-engine/vespa/pull/3668

Merge with https://git.corp.yahoo.com/vespa/hosted/pull/4871

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/Application.java [hunk 1]
```diff
@@ -129,8 +130,8 @@ public Application withProjectId(long projectId) {
         return new Application(id, deploymentSpec, validationOverrides, deployments, deploymentJobs.withProjectId(projectId), deploying, outstandingChange);
     }
 
-    public Application withJiraIssueId(Optional<String> jiraIssueId) {
-        return new Application(id, deploymentSpec, validationOverrides, deployments, deploymentJobs.withJiraIssueId(jiraIssueId), deploying, outstandingChange);
+    public Application withIssueId(IssueId issueId) {
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/Application.java]: Why not `with(IssueId issue)`? Existing code overloads when possible.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/api/Tenant.java [hunk 2]
```diff
@@ -29,24 +29,25 @@ public Tenant(TenantId id, Optional<UserGroup> userGroup, Optional<Property> pro
 
     public Tenant(TenantId id, Optional<UserGroup> userGroup, Optional<Property> property, Optional<AthenzDomain> athenzDomain, Optional<PropertyId> propertyId) {
         if (id.isUser()) {
-            require(!userGroup.isPresent(), "User tenant '%s' cannot have a user group.", id);
-            require(!property.isPresent(), "User tenant '%s' cannot have a property.", id);
-            require(!propertyId.isPresent(), "User tenant '%s' cannot have a property ID.", id);
-            require(!athenzDomain.isPresent(), "User tenant '%s' cannot have an athens domain.", id);
+            require( ! userGroup.isPresent(), "User tenant '%s' cannot have a user group.", id);
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/api/Tenant.java]: Never understood the reason for doing this, but OK.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/api/Tenant.java [hunk 3]
```diff
@@ -29,24 +29,25 @@ public Tenant(TenantId id, Optional<UserGroup> userGroup, Optional<Property> pro
...
-            require(!athenzDomain.isPresent(), "OpsDb tenant '%s' cannot have an athens domain.", id);
+            require(   property.isPresent(), "OpsDB tenant '%s' must have a property.", id);
+            require(   userGroup.isPresent(), "OpsDb tenant '%s' must have a user group.", id);
+            require( ! athenzDomain.isPresent(), "OpsDb tenant '%s' cannot have an athens domain.", id);
         }
         this.id = id;
         this.userGroup = userGroup;
         this.property = property;
         this.athenzDomain = athenzDomain;
         this.propertyId = propertyId; // TODO: Check validity after TODO@14. OpsDb tenants have this set in Sherpa, while athens tenants do not.
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/api/Tenant.java]: Should remove references to internal services.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentJobs.java [hunk 4]
```diff
@@ -75,21 +76,21 @@ public DeploymentJobs withTriggering(JobType jobType,
...
     }
 
     public DeploymentJobs withProjectId(long projectId) {
-        return new DeploymentJobs(Optional.of(projectId), status, jiraIssueId);
+        return new DeploymentJobs(Optional.of(projectId), status, issueId);
     }
 
-    public DeploymentJobs withJiraIssueId(Optional<String> jiraIssueId) {
-        return new DeploymentJobs(projectId, status, jiraIssueId);
+    public DeploymentJobs withIssueId(IssueId issueId) {
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentJobs.java]: Now that it's a proper type, why not `with(IssueId issueId)`?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 5]
```diff
@@ -164,6 +165,7 @@ private HttpResponse handleGET(HttpRequest request) {
         if (path.matches("/application/v4/property")) return properties();
         if (path.matches("/application/v4/cookiefreshness")) return cookieFreshness(request);
         if (path.matches("/application/v4/tenant/{tenant}")) return tenant(path.get("tenant"), request);
+        if (path.matches("/application/v4/tenant/{tenant}/property")) return property(path.get("tenant"));
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: As discussed offline: I suggest adding this to the tenant response and removing this path.

Also, what is the reason for exposing this information in the API?

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/application/responses/property-info.json [hunk 6]
```diff
@@ -0,0 +1,13 @@
+{
+    "propertyUri": "www.properties.com/1234",
+    "contactsUri": "www.contacts.com/1234",
+    "issueCreationUri": "www.issues.com/1234",
```
Comment (mpolden) [inline controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/application/responses/property-info.json]: Use fake domains in tests.

File: vespajlib/src/main/java/com/yahoo/data/access/slime/SlimeAdapter.java [hunk 7]
```diff
@@ -126,31 +131,27 @@ public String asString(String defaultValue) {
...
     public void traverse(final com.yahoo.data.access.ObjectTraverser ot) {
-        inspector.traverse(new com.yahoo.slime.ObjectTraverser() {
-                public void field(String name, com.yahoo.slime.Inspector inspector) { ot.field(name, new SlimeAdapter(inspector)); }
-            });
+        inspector.traverse((ObjectTraverser) (name, inspector) -> ot.field(name, new SlimeAdapter(inspector)));
     }
     public com.yahoo.data.access.Inspector entry(int idx) { return new SlimeAdapter(inspector.entry(idx)); }
     public com.yahoo.data.access.Inspector field(String name) { return new SlimeAdapter(inspector.field(name)); }
     public Iterable<com.yahoo.data.access.Inspector> entries() {
-        final List<com.yahoo.data.access.Inspector> list = new ArrayList<>();
```
Comment (mpolden) [inline vespajlib/src/main/java/com/yahoo/data/access/slime/SlimeAdapter.java]: Unrelated and also seems to slightly change behavior? Please remove this and put it in a separate PR to a module owner.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentIssueReporter.java [hunk 8]
```diff
@@ -8,52 +8,35 @@
...
-
-import static com.yahoo.vespa.hosted.controller.api.integration.Contacts.Category.admin;
-import static com.yahoo.vespa.hosted.controller.api.integration.Issues.IssueInfo.Status.done;
+import java.util.logging.Level;
 
 /**
- * Maintenance job which creates Jira issues for tenants when they have jobs which fails continuously 
- * and escalates issues which are not handled.
+ * Maintenance job which files issues for tenants when they have jobs which fails continuously
+ * and escalates issues which are not handled in a timely manner.
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentIssueReporter.java]: Quite the mouthful. Consider rephrasing :-)

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 9]
```diff
@@ -990,6 +1006,21 @@ private Slime toSlime(Tenant tenant, HttpRequest request, boolean listApplicatio
         return slime;
     }
 
+    private Slime toSlime(URI propertyUri, URI contactsUri, URI issueCreationUri, List<? extends List<? extends User>> contacts) {
+        Slime slime = new Slime();
+        Cursor root = slime.setObject();
+        root.setString("propertyUri", propertyUri.toString());
+        root.setString("contactsUri", contactsUri.toString());
+        root.setString("issueCreationUri", issueCreationUri.toString());
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: "Uri" -> "Url" in all these fields. This is how we name other URL fields as well. We sometimes confuse the terms as we always use `URI` to represent URLs as the `URL` class has strange behavior.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java [hunk 10]
```diff
@@ -990,6 +1006,21 @@ private Slime toSlime(Tenant tenant, HttpRequest request, boolean listApplicatio
...
+        Cursor lists = root.setArray("contacts");
+        for (List<? extends User> contactList : contacts) {
+            Cursor list = lists.addArray();
+            for (User contact : contactList)
+                list.addString(contact.toString());
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: Should use `username()` to avoid unintended consequences of changing `toString`.

**Review by mpolden — CHANGES_REQUESTED** [review]
Looks good overall, but I had a few comments.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/Organization.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+     */
+    IssueId file(Issue issue);
+
+    /**
+     * Returns the ID of this issue, if it exists and is open, based on a similarity search.
+     *
+     * @param issue The issue to search for; relevant fields are the summary and the owner (propertyId).
+     * @return ID of the issue, if it is found.
+     */
+    Optional<IssueId> findBySimilarity(Issue issue);
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/Organization.java]: I think `findSimilar` implies the argument itself is not to be returned, but here that is the whole point, so I prefer the current naming.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/Organization.java [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    /**
+     * Returns a nested list where the entries have increasing rank, and where each entry is
+     * a list of the users of that rank, by decreasing relevance.
+     *
+     * @param propertyId ID of the property for which to list contacts.
+     * @return A sorted, nested, reverse sorted list of contacts.
+     */
+    List<? extends List<? extends User>> contactsFor(PropertyId propertyId);
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/Organization.java]: Yes, subclassing User, which is useful when you want to sort them by rank and relevance, necessitates this.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/Application.java (hunk 1 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/Application.java]: Yes, good idea. It had to be named when it was not a custom class; not so anymore!

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentJobs.java (hunk 4 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentJobs.java]: Yes, will change that.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/api/Tenant.java (hunk 3 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/api/Tenant.java]: Yes.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentIssueReporter.java (hunk 8 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentIssueReporter.java]: Will do.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java (hunk 5 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: It was requested by the dashboard team.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java (hunk 9 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: Sure. And you're right, of course. Why not be specific when you can?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java (hunk 10 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: This is only for display in a table somewhere, and I wanted exactly the toString, since it includes the type of contact, in the subclass of User which we end up using in our implementation. 

I guess I could add a `displayName()` instead, then?

File: vespajlib/src/main/java/com/yahoo/data/access/slime/SlimeAdapter.java (hunk 7 above)
Comment (jonmv) [reply vespajlib/src/main/java/com/yahoo/data/access/slime/SlimeAdapter.java]: Whoo, where did that come from!? I'll remove it.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/Issue.java [hunk 13]
```diff
@@ -0,0 +1,68 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.controller.api.integration.organization;
+
+import com.yahoo.vespa.hosted.controller.api.identifiers.PropertyId;
+
+import java.util.Objects;
+import java.util.Optional;
+
+public class Issue {
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/Issue.java]: Done.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/Issue.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    public Issue append(String appendage) {
+        return new Issue(summary, description + appendage, label, assignee, propertyId);
+    }
+
+    public Issue withLabel(String label) {
+        return new Issue(summary, description, label, assignee, propertyId);
+    }
+    public Issue withAssignee(User assignee) {
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/Issue.java]: Fixed.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/IssueId.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.controller.api.integration.organization;
+
+import java.util.Objects;
+
+/**
+ * @author jvenstad
+ */
+public class IssueId {
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/IssueId.java]: Done.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/MockOrganization.java [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    @Override
+    public List<? extends List<? extends User>> contactsFor(PropertyId propertyId) {
+        return properties.get(propertyId).contacts;
+    }
+
+    @Override
+    public URI issueCreationUri(PropertyId propertyId) {
+        return URI.create("www.issues.com/" + propertyId.id());
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/MockOrganization.java]: Sure.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/MockOrganization.java [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    @Override
+    public URI contactsUri(PropertyId propertyId) {
+        return URI.create("www.contacts.com/" + propertyId.id());
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/MockOrganization.java]: Yes master.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/MockOrganization.java [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    @Override
+    public URI propertyUri(PropertyId propertyId) {
+        return URI.create("www.properties.com/" + propertyId.id());
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/MockOrganization.java]: Right away!

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/MockOrganization.java [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+            this.issue = issue;
+
+            updated = clock.instant();
+            open = true;
+            assignee = issue.assignee().orElse(properties.get(issue.propertyId()).defaultAssignee);
+        }
+
+    }
+
+
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/MockOrganization.java]: I like to put an extra newline around classes.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/MockOrganization.java [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+    public void addProperty(PropertyId propertyId) {
+        properties.put(propertyId, new PropertyInfo());
+    }
+
+    private void touch(IssueId issueId) {
+        issues.get(issueId).updated = clock.instant();
+    }
+
+
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/MockOrganization.java]: Same as below o_O

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/Organization.java [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+
+import com.yahoo.vespa.hosted.controller.api.identifiers.PropertyId;
+
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+
+public interface Organization {
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/Organization.java]: Done.

File: controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/User.java [hunk 22]
```diff
@@ -0,0 +1,43 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.controller.api.integration.organization;
+
+import java.util.Objects;
+
+public class User {
```
[the remark this answers is not in the document]
Comment (jonmv) [reply controller-api/src/main/java/com/yahoo/vespa/hosted/controller/api/integration/organization/User.java]: Done.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentJobs.java (hunk 4 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/DeploymentJobs.java]: Done.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentIssueReporter.java (hunk 8 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/DeploymentIssueReporter.java]: In a separate PR.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java (hunk 5 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: I changed my mind on this; 

The extra stuff displayed as propertyInfo requires lookups in OpsDB and JIRA, and I don't want to screw up all the 6 different tenant-renderer-paths when one of these external services fail to reply, time out, or whatever, so I think it's better to put it at a separate location.

Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: Perhaps this info should be stored in Tenant, and refreshed periodically by maintainers, as we do with other such information, but I think that's work for later. 

Tenant needs some work, anyway.

Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/restapi/application/ApplicationApiHandler.java]: I'll put the stuff on the tenant path anyway, for now, and make sure loading times are low later, if needed.

File: controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/application/responses/property-info.json (hunk 6 above)
Comment (jonmv) [reply controller-server/src/test/java/com/yahoo/vespa/hosted/controller/restapi/application/responses/property-info.json]: Done.

**Review by mpolden — APPROVED** [review]
